### PR TITLE
Remove `StreamingConnectionIdleTimeout` kubelet option

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
The option is not used any more and the timeout is handled by the container runtime.

Removing the CLI because the kubelet already states:
```
Flag --streaming-connection-idle-timeout has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
```

#### Which issue(s) this PR fixes:

None

#### Special notes for your reviewer:
cc @kwilczynski 
#### Does this PR introduce a user-facing change?

```release-note
Removed `StreamingConnectionIdleTimeout`/`--streaming-connection-idle-timeout` kubelet option because it is not used any more.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
None
```

### CODE REVIEW

**Comment by aojea** [comment]
if you remove the flags, systems that are deployed today setting the flag, are going to fail once they upgrade

@liggitt you have always more experience with all these deprecations, any suggestions?

**Comment by kwilczynski** [comment]
> if you remove the flags, systems that are deployed today setting the flag, are going to fail to run once they upgrade

This is why @rphillips suggested deprecating the field first and then eventually removing it.

> @liggitt you have always more experience with all these deprecations, any suggestions?

I suppose, we would issue a warning if this is set, so that people are aware that this is a no-op, especially if they had hoped this works and sets a timeout.

**Comment by saschagrunert** [comment]
@aojea @kwilczynski as mentioned in the top of the PR, the CLI is already deprecated and the kubelet prints the following warning if set:

```
> kubelet --streaming-connection-idle-timeout 4s
Flag --streaming-connection-idle-timeout has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
```

**Comment by kwilczynski** [comment]
> @aojea @kwilczynski as mentioned in the top of the PR, the CLI is already deprecated and the kubelet prints the following warning if set:
> 
> ```
> > kubelet --streaming-connection-idle-timeout 4s
> Flag --streaming-connection-idle-timeout has been deprecated, This parameter should be set via the config file specified by the Kubelet's --config flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.
> ```

Correct. :smile: 

I think that @aojea meant that once we remove this field from the type, then it shows up for users in the logs while parsing the configuration file for kubelet, for example:

```
Nov 30 08:35:35 kwilczynski-dev-fgzxq-worker-a-tx6v9 hyperkube[1199324]: I1130 08:35:35.383196 1199324 codec.go:117] "Using lenient decoding as strict decoding failed" err="strict decoding error: unknown field \"streamingConnectionIdleTimeout\""
```

That said, kubelet started without issues, so perhaps it wouldn't impact any user. :+1:

**Comment by aojea** [comment]
if I have `kubelet --streaming-connection-idle-timeout 5` in my systemd, scripts, whatever, since this change it will fail to start

**Comment by kwilczynski** [comment]
> if I have `kubelet --streaming-connection-idle-timeout 5` in my systemd, scripts, whatever, since this change it will fail to start because the flag does not exist, don't know how many people or jobs have those set, but those are going to break until they remove the flag from their scripts

@aojea, this has been deprecated for a while now.

Nothing is wrong with removing things that have been deprecated for a while. Even if this means that some users would notice failure and have to upgrade their scripts - a hard failure here is better since you get the feedback quickly (and know what to update/change) rather than a silent error (for lack of a better word) since users keep setting this up, but it does not do anything at the moment which I think is far worse.

**Comment by aojea** [comment]
> Rule #1: API elements may only be removed by incrementing the version of the API group.

https://kubernetes.io/docs/reference/using-api/deprecation-policy/

File: cmd/kubelet/app/options/options.go [hunk 1]
```diff
@@ -434,7 +435,9 @@ func AddKubeletConfigFlags(mainfs *pflag.FlagSet, c *kubeletconfig.KubeletConfig
...
-	fs.DurationVar(&c.StreamingConnectionIdleTimeout.Duration, "streaming-connection-idle-timeout", c.StreamingConnectionIdleTimeout.Duration, "Maximum time a streaming connection can be idle before the connection is automatically closed. 0 indicates no timeout. Example: '5m'. Note: All connections to the kubelet server have a maximum duration of 4 hours.")
+	dummyDuration := 4 * time.Hour
+	fs.DurationVar(&dummyDuration, "streaming-connection-idle-timeout", dummyDuration, "Maximum time a streaming connection can be idle before the connection is automatically closed. 0 indicates no timeout. Example: '5m'. Note: All connections to the kubelet server have a maximum duration of 4 hours.")
+	fs.MarkDeprecated("streaming-connection-idle-timeout", "This flag has no effect and will be removed in a future version.") //nolint:errcheck
```
Comment (aojea) [inline cmd/kubelet/app/options/options.go:440]: I was checking previous precedent , it seems  is ok to remove after some time announcing the deprecation https://github.com/kubernetes/kubernetes/pull/114446/commits/556d3fc2b7b4dfd1d94d3039909ddb42d88f7c52

I defer to node owners to decide how to better do it, but I suggest to keep at least one release of warning

maybe 
```suggestion
	fs.MarkDeprecated("streaming-connection-idle-timeout", "This flag has no effect and will be removed in v1.31.") //nolint:errcheck
```
??

**Comment by aojea** [comment]
> Nothing is wrong with removing things that have been deprecated for a while

> [...]
> 
> > > Nothing is wrong with removing things that have been deprecated for a while
> > 
> > 
> > that is not true, do you have more examples, we have a strong commitment with not breaking users,
> 
> How does this apply to a functionality that is currently broken and non-functional? This command-line option does not do anything at the moment, nor does the kubelet configuration option.
> 
> My suggestion here is about something that does not do anything any more. There is no valid way to set an idle timeout for streaming connections using this command-line switch (also true for kubelet configuration option).
> 
> We had users opening support cases against OpenShift who set this option and were surprised it does not work.
> 
> Should this be a genuine piece of functionality we are removing, then I wouldn't say anything. However, this is not the case here. This is more like a leftover from the dockershim removal time that was never removed since following deprecation a long time ago.
> 
> That said, given that the policy is clear about the removal of deprecated functionality and that you argue against removing it, then it's best we follow the guidelines here - yours and the policy. 👍

@kwilczynski let me clarify , I think that removing it is ok, I'm just arguing about how to do it, IIUIC this functionality does not work for the long time so the API change is a noop.

However, if you remove the flag without previous announcement, if you have clusters with that flag set at the upgrade they will break, so I think that at least we should warn about this and add a release note about it so people prepare in advance.

**Comment by aojea** [comment]
an example of potential end users that will have their cluster broken because we remove a flag without notice https://grep.app/search?q=--streaming-connection-idle-timeout

File: cmd/kubelet/app/options/options.go (hunk 1 above)
Comment (saschagrunert) [reply cmd/kubelet/app/options/options.go]: Changed as suggested.

**Comment by kwilczynski** [comment]
[...]
> > That said, given that the policy is clear about the removal of deprecated functionality and that you argue against removing it, then it's best we follow the guidelines here - yours and the policy. 👍
> 
> @kwilczynski let me clarify , I think that removing it is ok, I'm just arguing about how to do it, IIUIC this functionality does not work for the long time so the API change is a noop.

Yeah. That is why I was so eager to rid of it. Perhaps... a little too eager. :smile: 

> However, if you remove the flag without previous announcement, if you have clusters with that flag set at the upgrade they will break, so I think that at least we should warn about this and add a release note about it so people prepare in advance.

Fully agree. :+1: 

I checked the example you linked later—thank you for that, by the way—and it would be only prudent to ensure that users are both aware and can prepare. Many of these security compliance benchmarks rely on this more than I anticipated, so it seems.

**Pull request closed without merging**